全体のコード

```go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"

	"golang.org/x/sync/errgroup"
)

func main() {
	if err := run(context.Background()); err != nil {
		log.Printf("failed to terminate server: %v", err)
	}
}

func run(ctx context.Context) error {
	s := &http.Server{
		Addr: ":18080",
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintf(w, "Hello, %s!", r.URL.Path[1:])
		}),
	}
	eg, ctx := errgroup.WithContext(ctx)
	// 別ゴルーチンでHTTPサーバーを起動する
	eg.Go(func() error {
		// http.ErrServerClosed は
		// http.Server.Shutdown() が正常に終了したことを示すので異常ではない。
		if err := s.ListenAndServe(); err != nil &&
			err != http.ErrServerClosed {
			log.Printf("failed to close: %+v", err)
			return err
		}
		return nil
	})
	// チャネルからの通知（終了通知）を待機する
	<-ctx.Done()
	if err := s.Shutdown(context.Background()); err != nil {
		log.Printf("failed to shutdown: %+v", err)
	}
	// Goメソッドで起動した別ゴルーチンの終了を待つ。
	return eg.Wait()
}

```




## main関数
context.Context型の値を引数に渡すことで
- os.Exitを使わない
- エラーを返すようにすることで、テストできるようになる

```go
func main() {
	if err := run(context.Background()); err != nil {
		log.Printf("failed to terminate server: %v", err)
	}
}
```

## run関数
関数外部からサーバーを中断を受け取った際に、サーバーを中断可能にする
http.ListenAndServe関数を使っていると、サーバーを起動した後に、サーバーを中断することができない
そのため、http.Server型の値を作成し、サーバーを起動する

`http.Server`型の値を作成することで
- `ListenAndServe`メソッド: サーバーを起動し、HTTPリクエストを受信・処理（Handler関数に渡す）
- `Shutdown`メソッド: サーバーを中断
- タイムアウトも設定できたりする


```go
	s := &http.Server{
		Addr: ":18080",
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintf(w, "Hello, %s!", r.URL.Path[1:])
		}),
	}

    if err := s.ListenAndServe(); err != nil &&
        err != http.ErrServerClosed {
        log.Printf("failed to close: %+v", err)
        return err
    }

```



## グレースフルシャットダウンの実装
`golang.org/x/sync/errgroup`パッケージ
- 複数のゴルーチンを一括で管理できる
- ゴルーチンの終了を待つことができる

```go
    // errgroup を使用して複数のゴルーチンのエラーハンドリングを統一管理
    // いずれかのゴルーチンでエラーが発生すると、コンテキストがキャンセルされ、他のゴルーチンにも停止信号が送られます
	eg, ctx := errgroup.WithContext(ctx)
	// 別ゴルーチンでHTTPサーバーを起動する
	eg.Go(func() error {
		// http.ErrServerClosed は正常なシャットダウン時に返されます。エラーではありません。
		if err := s.ListenAndServe(); err != nil &&
			err != http.ErrServerClosed {
			log.Printf("failed to close: %+v", err)
			return err
		}
		return nil
	})
	// コンテキストがキャンセルされるのを待つ
	<-ctx.Done()
    // グレースフルシャットダウン
    // 新しいリクエストの受付を停止し、既存の接続の処理完了を待ってからサーバーを停止
	if err := s.Shutdown(context.Background()); err != nil {
		log.Printf("failed to shutdown: %+v", err)
	}

    //errgroupで起動したすべてのゴルーチンの終了を待機
	return eg.Wait()
```


## run関数のテスト
- 期待通りにHTTPサーバーが起動しているか
- テストコードから意図通りに終了するか
```go
package main

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"testing"

	"golang.org/x/sync/errgroup"
)

func TestRun(t *testing.T) {
    // キャンセル可能なコンテキストを作成
	ctx, cancel := context.WithCancel(context.Background())
	eg, ctx := errgroup.WithContext(ctx)

    // 別ゴルーチンでテスト対象のrun関数を実行
	eg.Go(func() error {
		return run(ctx)
	})

    // テスト用のHTTPリクエストを送信
	in := "message"
	rsp, err := http.Get("http://localhost:18080/" + in)
	if err != nil {
		t.Errorf("failed to get: %+v", err)
	}
	defer rsp.Body.Close()
	got, err := io.ReadAll(rsp.Body)
	if err != nil {
		t.Fatalf("failed to read body: %v", err)
	}
	// HTTPサーバーの戻り値を検証する（文字列かどうか）
	want := fmt.Sprintf("Hello, %s!", in)
	if string(got) != want {
		t.Errorf("want %q, but got %q", want, got)
	}
	// run関数に終了通知を送信する。
	cancel()
	// run関数の戻り値を検証する
	if err := eg.Wait(); err != nil {
		t.Fatal(err)
	}
}
```

実行コマンド
```
go test -v ./...
```
